R/PipeOpColRoles.R

```diff
@@ -57,52 +74,77 @@ PipeOpColRoles = R6Class("PipeOpColRoles",
       ps = ps(
-        # named list, each entry with a vector of roles
+        # named list, each entry with a vector of roles, names are columns
         new_role = p_uty(
           tags = c("train", "predict"),
           custom_check = crate(function(x) {
             first_check = check_list(x, types = "character", any.missing = FALSE, min.len = 1L, names = "named")
-            # return the error directly if this failed
-            if (is.character(first_check)) {
-              return(first_check)
-            }
-            # changing anything target related is not supported
-            # a value of "character()" will lead to the column being dropped
+            # Return the error directly if this failed
+            if (is.character(first_check)) return(first_check)
+            # Changing anything target related is not supported.
+            # A value of "character()" is accepted.
             all_col_roles = unique(unlist(mlr3::mlr_reflections$task_col_roles))
             check_subset(unlist(x), all_col_roles[all_col_roles != "target"])
           }, .parent = topenv())
+        ),
+        # named list, each with a vector of columns, names are column roles
+        new_role_direct = p_uty(
+          tags = c("train", "predict"),
+          custom_check = crate(function(x) {
+            first_check = check_list(x, types = "character", any.missing = FALSE, min.len = 1L, unique = TRUE, names = "named", null.ok = TRUE)
+            # Return the error directly if this failed
+            if (is.character(first_check)) return(first_check)
+            # Changing anything target related is not supported.
+            # A value of "character()" is accepted.
+            all_col_roles = unique(unlist(mlr3::mlr_reflections$task_col_roles))
+            check_subset(names(x), all_col_roles[all_col_roles != "target"])
+          }, .parent = topenv())
         )
       )
       super$initialize(id, param_set = ps, param_vals = param_vals, can_subset_cols = FALSE)
     }
   ),
   private = list(
     .transform = function(task) {
-      new_role = self$param_set$values$new_role
+      new_role = self$param_set$values[["new_role"]]
+      new_role_direct = self$param_set$values[["new_role_direct"]]
 
-      if (is.null(new_role)) {
+      if (is.null(new_role) && is.null(new_role_direct)) {
         return(task)  # early exit
       }
+      if (!is.null(new_role) && !is.null(new_role_direct)) {
+        stop("Both parameters, 'new_role' and 'new_role_direct', are set. Provide only one parameter at a time.")
+      }
 
-      new_role_names = names(new_role)
-      ids = task$col_info$id
-      ids = ids[ids != "..row_ids"]
-      # names of "new_role" must be a subset of the column names of the task
-      assert_subset(new_role_names, choices = ids, empty.ok = FALSE)
+      # Get the columns for which we change roles
+      # Replace NULLs with character(0)
+      if (!is.null(new_role)) {
```

Comment:
convert new_role to new_role_direct or new_role_direct to new_role and have only a single code path below. See which one of these conversions make the rest of the function more elegant.